presto-native-execution/presto_cpp/main/expression/RowExpressionEvaluator.cpp

```diff
@@ -0,0 +1,419 @@
+      auto expr = compileExpression(specialFormExpr);
+      if (auto constantExpr =
+              std::dynamic_pointer_cast<const exec::ConstantExpr>(expr)) {
+        if (constantExpr->value()->isNullAt(0)) {
+          return getConstantExpression(constantExpr);
+        }
+      }
+      break;
+    }
+    case protocol::Form::AND: {
+      auto left = specialFormExpr->arguments[0];
+      auto right = specialFormExpr->arguments[1];
+      auto lexpr = compileExpression(left);
+      bool isLeftNull;
+
+      if (auto constantExpr =
+              std::dynamic_pointer_cast<const exec::ConstantExpr>(lexpr)) {
+        isLeftNull = constantExpr->value()->isNullAt(0);
+        if (!isLeftNull) {
+          if (auto constVector =
+                  constantExpr->value()->as<ConstantVector<bool>>()) {
+            if (constVector->valueAt(0) == false) {
+              return getConstantExpression(constantExpr);
+            } else {
+              return right;
+            }
+          }
+        }
+      }
+
+      auto rexpr = compileExpression(right);
+      if (auto constantExpr =
+              std::dynamic_pointer_cast<const exec::ConstantExpr>(rexpr)) {
+        if (isLeftNull && constantExpr->value()->isNullAt(0)) {
+          return getConstantExpression(constantExpr);
+        }
+        if (auto constVector =
+                constantExpr->value()->as<ConstantVector<bool>>()) {
+          if (constVector->valueAt(0) == true) {
+            return left;
+          }
+          return right;
+        }
+      }
+      break;
+    }
+    case protocol::Form::OR: {
+      auto left = specialFormExpr->arguments[0];
+      auto right = specialFormExpr->arguments[1];
+      auto lexpr = compileExpression(left);
+      bool isLeftNull;
+
+      if (auto constantExpr =
+              std::dynamic_pointer_cast<const exec::ConstantExpr>(lexpr)) {
+        isLeftNull = constantExpr->value()->isNullAt(0);
+        if (!isLeftNull) {
+          if (auto constVector =
+                  constantExpr->value()->as<ConstantVector<bool>>()) {
+            if (constVector->valueAt(0) == true) {
+              return getConstantExpression(constantExpr);
+            }
+            return right;
+          }
+        }
+      }
+
+      auto rexpr = compileExpression(right);
+      if (auto constantExpr =
+              std::dynamic_pointer_cast<const exec::ConstantExpr>(rexpr)) {
+        if (isLeftNull && constantExpr->value()->isNullAt(0)) {
+          return getConstantExpression(constantExpr);
+        }
+        if (auto constVector =
+                constantExpr->value()->as<ConstantVector<bool>>()) {
+          if (constVector->valueAt(0) == false) {
+            return left;
+          }
+          return right;
+        }
+      }
+      break;
+    }
+    case protocol::Form::IN:
+    case protocol::Form::DEREFERENCE:
+    case protocol::Form::SWITCH:
+    case protocol::Form::WHEN:
+    case protocol::Form::COALESCE:
```

Comment:
There might be special case to filter out NULL values in COALESCE. If yes, then just add a TODO comment here.